src/caffe/net.cpp

```diff
@@ -650,6 +650,61 @@ void Net<Dtype>::ShareTrainedLayersWith(const Net* other) {
 }
 
 template <typename Dtype>
+void Net<Dtype>::ShareBlobsWith(const Net* other) {
+  int num_source_blobs = other->blobs().size();
+  for (int i = 0; i < num_source_blobs; ++i) {
+    const string& source_blob_name = other->blob_names()[i];
+    if (
+      std::find(
+        other->input_blob_indices().begin(),
+        other->input_blob_indices().end(), i) !=
+      other->input_blob_indices().end()) {
+      // input blob cannot be shared
+      LOG(INFO) << "Ignoring source blob " << source_blob_name;
+      continue;
+    }
+    int target_blob_id = 0;
+    while (target_blob_id != blob_names_.size() &&
+      blob_names_[target_blob_id] != source_blob_name) {
+      ++target_blob_id;
+    }
+    if (target_blob_id == blob_names_.size() ||
+      std::find(
+        net_input_blob_indices_.begin(),
+        net_input_blob_indices_.end(), target_blob_id) !=
+      net_input_blob_indices_.end()) {
+      LOG(INFO) << "Ignoring source blob " << source_blob_name;
+      continue;
+    }
+    LOG(INFO) << "Sharing source blob " << source_blob_name;
+    // We do not care blob shapes because every call of `Forward` method
+    // reshapes output blobs into right shapes, except input blobs.
+    blobs_[target_blob_id] = other->blobs()[i];
```

Comment:
Is this better than using Blob<Dtype>::ShareData() (and possibly ShareDiff)? E.g. 
```c++
blobs_[target_blob_id]->ShareData( *other->blobs()[i]));
```
Then you don't need UpdateBlobPointers because each net just keeps its own. I think this is more robust; ShareData() doesn't require that blob pointers be reassigned as in  UpdateBlobPointers. If you assign blob pointers instead of using ShareData, UpdateBlobPointers has to know all the places that blob pointers reside and keep them in sync. That introduces tight coupling to the rest of the implementation which is not robust but rather prone to breaking. 
Actually, if you do ShareData(),  I think you have to reshape before sharing because ShareData checks that the shape is the same.